src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -2335,6 +2336,54 @@ void MergeTreeData::stopOutdatedAndUnexpectedDataPartsLoadingTask()
     }
 }
 
+void MergeTreeData::prewarmMarkCache(ThreadPool & pool)
+{
+    if (!(*getSettings())[MergeTreeSetting::prewarm_mark_cache])
+        return;
+
+    auto * mark_cache = getContext()->getMarkCache().get();
+    if (!mark_cache)
+        return;
+
+    auto metadata_snaphost = getInMemoryMetadataPtr();
+    auto column_names = getColumnsToPrewarmMarks(*getSettings(), metadata_snaphost->getColumns().getAllPhysical());
+
+    if (column_names.empty())
+        return;
+
+    Stopwatch watch;
+    LOG_TRACE(log, "Prewarming mark cache");
+
+    auto data_parts = getDataPartsVectorForInternalUsage();
+
+    /// Prewarm mark cache firstly for the most fresh parts according
+    /// to time columns in partition key (if exists) and by modification time.
+
+    auto to_tuple = [](const auto & part)
+    {
+        return std::make_tuple(part->getMinMaxDate().second, part->getMinMaxTime().second, part->modification_time);
+    };
+
+    std::sort(data_parts.begin(), data_parts.end(), [&to_tuple](const auto & lhs, const auto & rhs)
+    {
+        return to_tuple(lhs) > to_tuple(rhs);
+    });
+
+    ThreadPoolCallbackRunnerLocal<void> runner(pool, "PrewarmMarks");
+
+    for (const auto & part : data_parts)
+    {
+        if (mark_cache->sizeInBytes() >= mark_cache->maxSizeInBytes() * 0.95)
```

Comment:
should a setting be added for this ratio to be changed if needed?

Reply:
Yes, we can add a setting.